Sources/Apollo/HTTPNetworkTransport.swift

```diff
@@ -1,11 +1,78 @@
+  /// Good for early-exiting if your user is not logged in, for example.
+  ///
+  /// - Parameters:
+  ///   - networkTransport: The network transport which wants to send a request
+  ///   - request: The request, BEFORE it has been modified by `willSend`
+  /// - Returns: True if the request should proceed, false if not.
+  func networkTransport(_ networkTransport: HTTPNetworkTransport, shouldSend request: URLRequest) -> Bool
+  
+  /// Called when a request is about to send. Allows last minute modification of any properties on the request,
+  ///
+  ///
+  /// - Parameters:
+  ///   - networkTransport: The network transport which is about to send a request
+  ///   - request: The request, as an `inout` variable for modification
+  func networkTransport(_ networkTransport: HTTPNetworkTransport, willSend request: inout URLRequest)
+}
+
+// MARK: -
+
+/// Methods which will be called after some kind of response has been received to a `URLSessionTask`.
+public protocol HTTPNetworkTransportTaskCompletedDelegate {
+  
+  /// A callback to allow hooking in URL session responses for things like logging and examining headers.
+  /// NOTE: This will call back on whatever thread the URL session calls back on, which is never the main thread. Call `DispatchQueue.main.async` before touching your UI!
+  ///
+  /// - Parameters:
+  ///   - networkTransport: The network transport that completed a task
+  ///   - request: The request which was completed by the task
+  ///   - data: [optional] Any data received. Passed through from `URLSession`.
+  ///   - response: [optional] Any response received. Passed through from `URLSession`.
+  ///   - error: [optional] Any error received. Passed through from `URLSession`.
+  func networkTransport(_ networkTransport: HTTPNetworkTransport,
+                        completedRawTaskForRequest request: URLRequest,
+                        withData data: Data?,
+                        response: URLResponse?,
+                        error: Error?)
+}
+
+// MARK: -
+
+public protocol HTTPNetworkTransportRetryDelegate: class {
+  
+  /// Called when an error has been received after a request has been sent to the server to see if an operation should be retried or not.
+  /// NOTE: Don't just call the `retryHandler` with `true` all the time, or you can potentially wind up in an infinite loop of errors
+  ///
+  /// - Parameters:
+  ///   - networkTransport: The network transport which received the error
+  ///   - error: The received error
+  ///   - request: The URLRequest which generated the error
+  ///   - response: [Optional] Any response received when the error was generated
+  ///   - retryHandler: A closure indicating whether the operation should be retried. Asyncrhonous to allow for re-authentication or other async operations to complete.
+  func networkTransport(_ networkTransport: HTTPNetworkTransport,
+                        receievedError error: Error,
```

Comment:
Small typo, should be `receivedError`.

Reply:
🤦‍♀️If i spell `received` right on the first try ONCE IN MY LIFE I will be happy.